src/server/bitops_family.cc

```diff
@@ -57,14 +65,88 @@ constexpr int32_t GetByteIndex(uint32_t offset) noexcept {
   return offset / OFFSET_FACTOR;
 }
 
-uint8_t GetByteValue(const std::string& str, uint32_t offset) {
+uint8_t GetByteValue(std::string_view str, uint32_t offset) {
   return static_cast<uint8_t>(str[GetByteIndex(offset)]);
 }
 
 constexpr bool CheckBitStatus(uint8_t byte, uint32_t offset) {
   return byte & (0x1 << offset);
 }
 
+constexpr std::uint8_t CountBitsRange(std::uint8_t byte, std::uint8_t from, uint8_t to) {
+  int count = 0;
+  for (int i = from; i < to; i++) {
+    count += CheckBitStatus(byte, GetNormalizedBitIndex(i));
+  }
+  return count;
+}
+
+constexpr std::uint8_t BitOnCount(std::uint8_t byte) {
+  return NIBBLE_LOOKUP[byte & 0x0F] + NIBBLE_LOOKUP[byte >> 4];
+}
+
+std::uint32_t CountSetAllBytes(std::string_view at, int32_t start, int32_t end) {
+  if (start >= end || start < 0 || end < 0) {
+    return 0;
+  }
+  end = std::min(end, static_cast<int32_t>(at.size()));  // don't overflow
+  std::uint32_t count = std::accumulate(
+      std::next(at.begin(), start), std::next(at.begin(), end), 0,
+      [](auto counter, auto ch) { return counter + BitOnCount(static_cast<std::uint8_t>(ch)); });
+  return count;
+}
+
+std::uint32_t CountAllBitsOn(std::string_view at, int32_t start, int32_t end) {
```

Comment:
Can you please add comments explaining what CountAllBitsOn does vs CountBitsOnBitOffset vs CountSetAllBytes.

it's a bit hard for me to follow

Reply:
Will add, but for this review, "CountAllBitsOn" is for counting on with bit number is the index and "CountBitsOnBitOffset" is for counting in bytes offset. I will try to think of better name, and would add comments